pydis_site/apps/api/serializers.py

```diff
@@ -249,15 +255,57 @@ class Meta:
         fields = ('id', 'name', 'colour', 'permissions', 'position')
 
 
-class UserSerializer(BulkSerializerMixin, ModelSerializer):
+class UserListSerializer(ListSerializer):
+    """List serializer for User model to handle bulk updates."""
+
+    def create(self, validated_data: list) -> list:
+        """Override create method to optimize django queries."""
+        present_users = User.objects.all()
+        new_users = []
+        present_user_ids = [user.id for user in present_users]
+
+        for user_dict in validated_data:
+            if user_dict["id"] in present_user_ids:
+                raise ValidationError({"id": "User already exists."})
+            new_users.append(User(**user_dict))
+
+        return User.objects.bulk_create(new_users)
+
+    def update(self, instance: QuerySet, validated_data: list) -> list:
+        """
+        Override update method to support bulk updates.
+
+        ref:https://www.django-rest-framework.org/api-guide/serializers/#customizing-multiple-update
+        """
+        instance_mapping = {user.id: user for user in instance}
+        data_mapping = {item['id']: item for item in validated_data}
+
+        updated = []
+        fields_to_update = set()
+        for user_id, data in data_mapping.items():
+            for key in data:
+                fields_to_update.add(key)
+            user = instance_mapping.get(user_id)
+            user.__dict__.update(data)
+            updated.append(user)
+
+        fields_to_update.remove("id")
+        User.objects.bulk_update(updated, fields_to_update)
+        return updated
+
+
+class UserSerializer(ModelSerializer):
     """A class providing (de-)serialization of `User` instances."""
 
+    id = IntegerField(min_value=0)
```

Comment:
Apparently it's because of this

> You will need to add an explicit id field to the instance serializer. The default implicitly-generated id field is marked as read_only. This causes it to be removed on updates. Once you declare it explicitly, it will be available in the list serializer's update method.

Worth adding a comment to specify why it's explicitly defined.